cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1367,11 +1393,34 @@ private boolean isRunning(Id.Program id) {
     return programStatus != null && !"STOPPED".equals(programStatus);
   }
 
+  private boolean isRunLevelActionAllowed(ProgramType type) {
+    // Run level actions only enabled for the Workflow and MapReduce
+    return EnumSet.of(ProgramType.WORKFLOW, ProgramType.MAPREDUCE).contains(type);
+  }
+
+  private boolean isConcurrentRunsDisabled(Id.Program id, Map<String, String> properties) {
+    if (!isRunLevelActionAllowed(id.getType())) {
+      return false;
+    }
+
+    return Boolean.parseBoolean(properties.get(ProgramOptionConstants.CONCURRENT_RUNS_DISABLED));
+  }
+
+  private AppFabricServiceStatus stop(Id.Program id) {
+    return stop(id, null);
+  }
+
   /**
    * Stops a Program.
    */
-  private AppFabricServiceStatus stop(Id.Program identifier) {
-    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(identifier);
+  private AppFabricServiceStatus stop(Id.Program identifier, String runId) {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo;
```

Comment:
I think it makes sense to abstract the next 6-7 lines into a method like `findRuntimeInfo(identifier, runId)`, since it'd be useful also when getting status by runId.